Dynamic Annotation Cleanup

rozmansi opened the pull request:
This PR aims to address as many suggestions reported in #623 as possible.

@lxn, please do not merge this PR before @zx2c4 takes a look at it. It is much easier for me to have this PR open and fix things inside it to make everyone happy, than reverting master back and forth.

- HRESULTs are no longer returned to the clients.
- All win.SOMETHING_UGLY constants have been redeclared in lxn/walk and created as enums. Like MessageBox does it.
- Constants were declared in accessibility.go to reduce clutter in windows.go.
- I am not sure about keeping Dynamic Annotation methods in the Window interface, but Dynamic Annotation properties are bound to each HWND like window text and other GDI properties. If you can't see them, it doesn't mean they are less important.
- All windows of the same windows group now share the same CLSID_AccPropServices object.
- OLE/COM initialization is no longer required/expected from the client. It is initialized in WindowGroup automatically. Like folder selection dialog or WebView control do. Unlike ProgressIndicator does not.

zx2c4 commented:
Looks like a step in the right direction, though seems like it'd be better to place these in accessibility.go instead. This also doesn't address the issue of bloating the interface but I'll let @lxn weigh in on that.

rozmansi commented:
Rebased this PR onto the new master and removed accessibility setters from Window interface.

> Looks like a step in the right direction, though seems like it'd be better to place these in accessibility.go instead. This also doesn't address the issue of bloating the interface but I'll let @lxn weigh in on that.

Detaching Dynamic Annotation from the window.go and windowbase.go would make things uglier. annotation.go would need to invent own scope for storing accPropServices object - something WindowGroup is already perfectly suitable to host. The accessibility property setters to be moved to accessibility.go would have to be changed from WindowBase methods to standalone functions taking HWND or something lxn/walk-friendly as a parameter...

Do you have any specific model in mind?

lxn commented:
You could do it like this:

```Go
// window.go
type WindowBase struct {
...
	acc *Accessibility
}

func (wb *WindowBase) Accessibility() *Accessibility {
	// Maybe create this on demand

	return wb.acc
}

// accessibility.go
type Accessibility struct {
	wb *WindowBase
}

func (acc *Accessibility) SetDescription(desc string) error {
...
}
```

rozmansi commented:
Thanks, @lxn.

May I leave the `accPropServices` and `OleInitialize()` in the `WindowGroup`? This is most resource-friendly, as the `accPropServices` may be reused by different windows of the same thread.

rozmansi commented:
Oh, and one more thing... I would like to make the accessibility first-class citizen. I.e. support setting DA properties in declarative subpackage too. Probably, I need to add accessiblity.go to the declarative subpackage akin to the base lxn/walk package, right?

lxn commented:
> May I leave the accPropServices and OleInitialize() in the WindowGroup?
Sure

> Probably, I need to add accessiblity.go to the declarative subpackage akin to the base lxn/walk package, right?
SGTM

rozmansi commented:
@zx2c4 you're welcome.
@lxn hold on with merging, while I add support in declarative.

zx2c4 commented on examples/actions/actions.go:
```diff
@@ -158,6 +161,9 @@ func main() {
 					}
 				},
+				Accessibility: Accessibility{
+					Help: "Toggles special mode",
```
Cool field. Out of curiosity, does this enable help tooltips? Or make that text be present when the question mark titlebar button is enabled? Or some other place?

rozmansi replied:
Yes, `Help` actually sets two window properties:
![image](https://user-images.githubusercontent.com/6889694/65751755-a7f12780-e10b-11e9-88a1-7615c62ab35b.png)

I'm not sure about who picks up the `Help` one on the screenshot, but the `IAccessible.Help` is picked up by screen readers.

lxn commented on declarative/accessibility.go:
```diff
@@ -0,0 +1,137 @@
+
+// Accessibility properties
+type Accessibility struct {
+	Accelerator   Property
```
Declarative struct fields shouldn't be of type `Property`, unless they have a `*WindowBase.Property()` counterpart. Please use the appropriate types like `string` for these.

rozmansi replied:
OK. Other than not being able to set control state explicitly to `AccStateNormal`, or clear some pre-set accessibility property (shortcuts are automatically created for static texts containing `'&'` character), I see no drawback to have them as `string`, `AccRole`, or `AccState`.

lxn commented on declarative/checkbox.go:
```diff
@@ -33,6 +33,7 @@ type CheckBox struct {
 	RightToLeftReading bool
 	ToolTipText        Property
 	Visible            Property
+	Accessibility      Accessibility
```
Please keep the fields sorted alphabetically.

rozmansi replied:
Ups. Thanks.

lxn commented:
Very nice, thanks!

jscholes commented:
@rozmansi Is being able to set AccStateNormal desirable?  For example, if I set AccStateExpanded or AccStateCollapsed to indicate that a button is a toggle, and then later change my mind for whatever reason, how would I remove those toggle semantics?  Likewise with flags like AccStateUnavailable which don't have corresponding inverse states?

rozmansi commented:
> @rozmansi Is being able to set AccStateNormal desirable? For example, if I set AccStateExpanded or AccStateCollapsed to indicate that a button is a toggle, and then later change my mind for whatever reason, how would I remove those toggle semantics? Likewise with flags like AccStateUnavailable which don't have corresponding inverse states?

That was my rationale to use `Property` type for `declarative.Accessibility` members initially - to be able to distinguish between "not specified" and "zero". In 580ebdf5528b3e0d3f0124a67246e011e67abb65, switching to `string` and enum property types you can no longer set accessibility properties to `""` or `0` using declarative. But, you can always call `control.Accessibility().SetAccelerator("")` or `control.Accessibility().SetState(walk.AccStateNormal)` to do so.
